### ISSUE TITLE
test-fs-util fails in Arch mkosi

### ISSUE THREAD

**Issue by daandemeyer** [systemd/systemd#15821]
**systemd version the issue has been seen with**
> master

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > Tests succeeds

**Unexpected behaviour you saw**
 > test-fs-util fails

**Steps to reproduce the problem**

`sudo mkosi --default .mkosi/mkosi.arch`

It fails immediately on `test_path_is_encrypted_one("/home", -1);` with `No such file or directory`. I did some more debugging and https://github.com/systemd/systemd/blob/32fcf399bfe7ff6e08ace48106567f8848a1b6e0/src/basic/btrfs-util.c#L318 fails on /dev/loop0p2 which doesn't exist. If I look in /dev after mkosi exits, /dev/loop0 is gone. This might be caused by the same issue as https://github.com/systemd/mkosi/issues/416.

I'm on Linux 5.6.12. 5.6.13 is in the testing repos, I'll report back if that version changes anything once it's released.

**Comment by jshap70** [comment]
I can confirm I was seeing this issue as well also on arch on the same kernel versions.

**Comment by DaanDeMeyer** [comment]
5.6.13 does not fix the issue. /dev/loop0 is still gone after mkosi finishes. `dmesg` doesn't show any errors or kernel stacktraces though.

**Comment by DaanDeMeyer** [comment]
So the dissapearing loop devices turned out to be completely normal. When allocating a loop device, we use `LOOP_CTL_GET_FREE` which returns one of the default allocated loop devices. When unreffing the loop device, we unconditionally call `LOOP_CTL_REMOVE` which removes the loop device even if it's one of the default ones. I find this extremely unintuitive, although its more because of `LOOP_CTL_GET_FREE` not indicating whether a new loop device was allocated or an existing one was reused. Maybe we can simply always use `LOOP_CTL_ADD` to avoid removing the default loop devices?

**Comment by DaanDeMeyer** [comment]
And with that out of the way I think we pretty easily find the cause of the test failure as well. /dev in the systemd-nspawn container doesn't contain any /dev/loop devices or their partitions which causes the `stat` failure I mentioned in the first post because it tries to stat /dev/loop0p2 which doesn't exist in the container's /dev filesystem.

I'm not entirely sure how to fix this. Do we mount the corresponding loop device into the container? Maybe we can avoid the loop device somehow? I guess this might be a more general problem where we simply don't have the required information available in a container. How do we make this a reliable test then though?

**Comment by poettering** [comment]
> So the dissapearing loop devices turned out to be completely normal. When allocating a loop device, we use LOOP_CTL_GET_FREE which returns one of the default allocated loop devices. When unreffing the loop device, we unconditionally call LOOP_CTL_REMOVE which removes the loop device even if it's one of the default ones. I find this extremely unintuitive, although its more because of LOOP_CTL_GET_FREE not indicating whether a new loop device was allocated or an existing one was reused. Maybe we can simply always use LOOP_CTL_ADD instead of LOOP_CTL_GET_FREE to avoid removing the default loop devices?

Loop block device allocation is really weird for historical reasons. You start out with zero devices, then you allocate and use one, but if you abort at an unconvenient place you might leave an allocated but unattached once around. Our code generally uses whatever is free regardless if allocated or not, so that we automatically recycle left-over devices. It's the only safe thing to do to not risk that unbounded numbers of block devices that are allocated but not used pile up.

**Comment by Werkov** [comment]
I came up against this test failure in the mkosi openSUSE container too (after ed9c0851e51b0a8c37fff3495c4abbb9ecdbe8d2).
I arrived at the (same) conclusion that this is because of private `/dev` mount, so (host) loop devices aren't visible but they're referenced (below)

```
ioctl(3, BTRFS_IOC_DEV_INFO, {devid=makedev(0, 0x1)} => {uuid=67046640-0d39-41e1-ab4a-bd17138cd55d, bytes_used=1631584256, total_bytes=3222253568, path="/dev/loop1p2"}) = 0
```

I'm not sure the `path_is_encrypted()` code should suppose that device node would always be accessible (this needn't to hold generally in the containers).

Or are there some more environment factors? (I wonder this didn't popped up in CI, but I guess they don't run in containers.)

**Comment by poettering** [comment]
My proposed fix is in #15845. Which is not a fix. It just accepts that we cannot make this work in a container

**Comment by poettering** [comment]
> Or are there some more environment factors? (I wonder this didn't popped up in CI, but I guess they don't run in containers.)

They do run in containers, but they don't use btrfs...

It's a limitation of btrfs if you so will that they return the devices as fs path instead of dev_t. That means we cannot determine everything we need from just /sys/, but need to check /dev/ first to convert a device path into a dev_t.
